lib/Analysis/DxilValueCache.cpp

```diff
@@ -180,6 +172,40 @@ Value *DxilValueCache::ProcessAndSimplify_PHI(Instruction *I, DominatorTree *DT)
   return Simplified;
 }
 
+Value *DxilValueCache::ProcessAndSimplify_Switch(Instruction *I, DominatorTree *DT) {
+  SwitchInst *Sw = cast<SwitchInst>(I);
+  BasicBlock *BB = Sw->getParent();
+
+  Value *Cond = TryGetCachedValue(Sw->getCondition());
+  if (IsUnreachable_(BB)) {
+    for (unsigned i = 0; i < Sw->getNumSuccessors(); i++) {
+      BasicBlock *Succ = Sw->getSuccessor(i);
+      if (Succ->getSinglePredecessor())
+        MarkUnreachable(Succ);
+    }
+  }
+  else if (isa<Constant>(Cond)) {
+    bool FoundCase = false;
+    for (auto Case : Sw->cases()) {
+      BasicBlock *Succ = Case.getCaseSuccessor();
+      if (Case.getCaseValue() == Cond) {
+        FoundCase = true;
+      }
+      else {
+        if (Succ->getSinglePredecessor())
+          MarkUnreachable(Succ);
+      }
+    }
+
+    BasicBlock *DefaultSucc = Sw->getDefaultDest();
+    if (FoundCase && DefaultSucc->getSinglePredecessor()) {
```

Comment:
Could the default dest be the same as the successor for the found case in some instance?  At which point it would be invalid to mark it unreachable here?

Reply:
If the switch has multiple cases branching to the same BB, BB->getSinglePredecessor() returns NULL, because it checks if BB has a single Use, which in this case it would have multiple, each being an operand of the switch statement. But this code is a bit confusing... I'll revise it.